lib/ecto/query/planner.ex

```diff
@@ -1707,6 +1707,11 @@ defmodule Ecto.Query.Planner do
     type!(kind, query, expr, ix, field, allow_virtuals?)
   end
 
+  defp field_type!(kind, query, expr, {{{:., _, [_, :count_alias!]}, _, [_, name]}, field}, allow_virtuals?) do
```

Comment:
```suggestion
  defp field_type!(kind, query, expr, {{{:., _, [__MODULE__, :count_alias!]}, _, [_, name]}, field}, allow_virtuals?) do
```

?

Reply:
If it doesn't pass with my comment, then change `calculate_named_binds` to this:

    unquote(__MODULE__).count_alias!(query, unquote(name))

Reply:
Wouldn't `__MODULE__` in `defp field_type!(... {{{:., _, [__MODULE__, :count_alias!]},  ...) do` refer to `Ecto.Query.Planner`?

I had to make both changes for it to pass but it still feels like a bandage over another issue...

The intention seems to be that some other part of the code would apply the `count_alias!` before this point, calculating the index and matching on the expression before this one.

However if you're happy with the changes so am I. I have some other tests that exercise the named select-type combination in a few ways if you're interested in them?

https://github.com/03juan/ecto_sql/blob/named-typed-column/test/ecto/adapters/postgres_test.exs#L112

Reply:
I agree that it is a bit awkward but I think doing the pre-processing after would be even more work. I will explore it a bit after merge!